sdk/servicebus/service-bus/src/core/linkEntity.ts

```diff
@@ -394,8 +409,14 @@ export abstract class LinkEntity<LinkT extends Receiver | AwaitableSender | Requ
    * @return {Promise<void>} Promise<void>
    */
   private async _negotiateClaim(setTokenRenewal?: boolean): Promise<void> {
+    log.error(`${this._logPrefix} negotiateclaim() has been called`);
+
     // Wait for the connectionContext to be ready to open the link.
-    await this._context.readyToOpenLink();
+    if (this._context.isConnectionClosing()) {
+      log.error(`${this._logPrefix} connection is closing, aborting negotiateClaim`);
+      throw new Error("Connection was closing, aborting negotiateClaim.");
```

Comment:
I see negotiateClaim being called from 2 places
- when link needs to be set up
- when we need to renew the claim

In the second case, we catch the error thrown by negotiateClaim, log it, but do not re-throw
In the first case, we allow the error to bubble up. Will this ever surface to the user? What is the user expected to do?

I see the connection closing case to be a "retryable" one. Will this error be translated to a retryable one?

Reply:
These are great questions and I think the answer comes down to "it depends" for a lot of it because of the varied locations it's called from.

For the initial connection, in at least most cases, we'll trigger retries. I've changed it to throw a MessagingError with .retryable explicitly set to guarantee this.

For token renewal (or detach) those can happen independently of a user operation so we don't have anything setup aside from logging. Perhaps we could change this in the future by requiring the operations to set up a callback so if the user happens to be using the link at that point we can report the error there. That's probably larger than this pull request but it's a possibility.

If we skip ahead a bit and just think about calls that are happening directly as a result of an operation - we should fall under the retry logic so the user doesn't have to do anything although it's possible, ultimately, for the error to actually reach the user once we've exhausted retries.

It might be nice to add in some sort of "retry is failing on the _last_ retry, transform the error into something else". Some thought required, I think.

Reply:
(haven't pushed this change yet, still testing)

Reply:
Pushed.

Reply:
Filed an issue to cover the work for when we renew the claim as part of #10921